src/ray/core_worker/core_worker.cc

```diff
@@ -1167,6 +1160,13 @@ void CoreWorker::Exit(
     const rpc::WorkerExitType exit_type,
     const std::string &detail,
     const std::shared_ptr<LocalMemoryBuffer> &creation_task_exception_pb_bytes) {
+  if (is_shutdown_) {
```

Comment:
Since `Shutdown` is only called in the `shutdown` callback within `Exit`, promising to execute `Exit` at most once also ensures that `Shutdown` is executed at most once.

Reply:
I don't think this is correct. `CoreWorker::Shutdown` is also called by `CoreWorkerProcessImpl::ShutdownDriver`, which is called in `CorwWorker.shutdown_driver` in `_raylet.pyx`.

https://github.com/ray-project/ray/blob/c6639d242f61d43cfc0db7f24628cc16e6af1bf7/src/ray/core_worker/core_worker_process.cc#L337-L349

https://github.com/ray-project/ray/blob/c6639d242f61d43cfc0db7f24628cc16e6af1bf7/python/ray/_raylet.pyx#L3047-L3055

Reply:
Should also add a test when `ray.shutdown` is called.

Reply:
Good catch! I only checked the call in `core_worker.cc`. I'm surprised there's another call to Shutdown outside of `core_worker.cc`. Maybe we should consider unifying the code paths for both the core worker driver and the worker shutdown process.

Reply:
I will use another flag instead of reusing `is_shutdown_`.

Reply:
> Should also add a test to check the case that ray.shutdown is called.

Is this still required if I don't reuse `is_shutdown_`?

Reply:
I add a new flag instead. Let me know if you think a test for `ray.shutdown` is still required.

https://github.com/ray-project/ray/pull/51582/commits/4e24d0ebd08ffadb35f609cf191e910d41762889

Reply:
@kevin85421 can you help me understand exactly the difference between `is_exit_` and `is_shutdown_` now?

Reply:
`is_shutdown_` is used to avoid executing `CoreWorker::Shutdown` twice.

`is_exit_` is used to avoid executing `CoreWorker::Exit` twice. `CoreWorker::Exit` also calls `CoreWorker::Shutdown`. However, `CoreWorker::Shutdown` also calls by functions outside `core_worker.cc` like `CoreWorkerProcessImpl::ShutdownDriver`.

Reply:
In the future, is there a way that we can better unify this shutdown sequence or is there a reason why we need separate public interfaces?

Reply:
I don't actually delve deeply into it, but I fully agree that we should have a single public function to terminate the core worker, whether it is a driver or a worker as I said https://github.com/ray-project/ray/pull/51582#discussion_r2008643130.

Reply:
We can leave `Exit` as the only one public interface to gracefully terminate a core worker process, and only `Exit` can call `Shutdown`.

Reply:
Can you take it as a follow-up to investigate and fix it if it isn't too hard?

Reply:
> Can you take it as a follow-up to investigate and fix it if it isn't too hard?

Sure! Create an issue to track the progress: https://github.com/ray-project/ray/issues/51642.